File: 283.move-zeroes.java
/*
 * @lc app=leetcode id=283 lang=java
 *
 * [283] Move Zeroes
 */

// @lc code=start
class Solution {
    public void moveZeroes(int[] nums) {
        int left = 0, right = 0;
        while (right != nums.length) {
            if (nums[right] != 0) {
                nums[left++] = nums[right];
            }
            right += 1;
        }
        while (left != nums.length) {
            nums[left++] = 0;
        }
    }
}
// @lc code=end

File: 283.move-zeroes.cpp
/*
 * @lc app=leetcode id=283 lang=cpp
 *
 * [283] Move Zeroes
 */

// @lc code=start

#include <vector>
using namespace std;

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        auto rptr = nums.begin(), wptr = nums.begin();
        while (rptr != nums.end()) {
            if (*rptr != 0) {
                *wptr = *rptr;
                wptr += 1;
            }
            rptr += 1;
        }
        std::fill(wptr, nums.end(), 0);
    }
};
// @lc code=end
